# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there is a `data/` folder with two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this notebook, we will combine these two data sets together into one pandas data frame.

# 1. Combining two datasets with the same columns

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [2]:
import pandas as pd

In [3]:
# TODO: import the pandas library

# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('data/rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('data/electricity_access_percent.csv', skiprows=4)

# TODO: remove the Unnamed: 62 column from each data set
df_rural.drop('Unnamed: 62', axis=1, inplace=True)
df_electricity.drop('Unnamed: 62', axis=1, inplace=True)

# TODO: combine the two data sets together using the concat method
# In other words, all of the rows of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.
df = pd.concat([df_rural, df_electricity])

In [4]:
df.head(2)

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   

   Indicator Code    1960    1961    1962   1963    1964   1965  ...    2008  \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.27  49.285  49.30  ...  56.217   
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.89  90.574  90.25  ...  76.054   

     2009    2010    2011    2012    2013    2014    2015    2016    2017  
0  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696  58.893  
1  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868  72.430  

[2 rows x 62 columns]

In [5]:
df.shape

(528, 62)

In [6]:
df_rural.shape

(264, 62)

In [7]:
df_electricity.shape

(264, 62)

# 2. Transform the datasets 
This task is more difficult.

Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [8]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value
df_rural_melt = pd.melt(df_rural,\
                        id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'],\
                       var_name = 'Year', value_name='Rural_Value')
df_electricity_melt = pd.melt(df_electricity,\
                              id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'],\
                             var_name='Year', value_name='Electricity_Value')

# TODO: drop any columns from the data frames that aren't needed
df_rural_melt.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)
df_electricity_melt.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year
df_merge = df_rural_melt.merge(df_electricity_melt, how='outer',\
                               on=['Country Name', 'Country Code', 'Year'])

# TODO: sort the results by country and then by year
df_combined = df_merge.sort_values(['Country Name', 'Year'])
df_combined

Country Name Country Code  Year  Rural_Value  Electricity_Value
1      Afghanistan          AFG  1960       91.779                NaN
265    Afghanistan          AFG  1961       91.492                NaN
529    Afghanistan          AFG  1962       91.195                NaN
793    Afghanistan          AFG  1963       90.890                NaN
1057   Afghanistan          AFG  1964       90.574                NaN
...            ...          ...   ...          ...                ...
14255     Zimbabwe          ZWE  2013       67.346          37.076813
14519     Zimbabwe          ZWE  2014       67.499          32.300000
14783     Zimbabwe          ZWE  2015       67.624          33.700000
15047     Zimbabwe          ZWE  2016       67.723          38.145138
15311     Zimbabwe          ZWE  2017       67.793                NaN

[15312 rows x 5 columns]